File: two-pointers/lint-533/lint-533.js
/**
 * @param nums: an integer array
 * @param target: An integer
 * @return: the difference between the sum and the target
 */
const twoSumClosest = function (nums, target) {
  nums.sort((a, b) => a - b);
  let left = 0;
  let right = nums.length - 1;
  let diff = Infinity;
  while (left < right) {
    const value = nums[left] + nums[right];
    if (value < target) {
      diff = Math.min(diff, target - value);
      left += 1;
    } else {
      diff = Math.min(diff, value - target);
      right -= 1;
    }
  }
  return diff;
};

File: two-pointers/lint-533/lint-533.py
# https://www.lintcode.com/problem/two-sum-closest-to-target/description?_from=ladder&&fromId=1

"""
Time: O(nlogn+n)->O(nlogn)
Space: O(1)
"""


class Solution:
    """
    @param nums: an integer array
    @param target: An integer
    @return: the difference between the sum and the target
    """

    def twoSumClosest(self, nums, target):
        # write your code here
        nums.sort()
        left, right = 0, len(nums) - 1
        diff = float("Inf")
        while left < right:
            value = nums[left] + nums[right]
            if value < target:
                diff = min(diff, target - value)
                left += 1
            else:
                diff = min(diff, value - target)
                right -= 1
        return diff
